### PR TITLE
Object Detection backbone freezing option

### PR DESCRIPTION
**Pull request by gabrielfruet**
## What has changed and why?

#615

This pull request adds support for freezing the backbone weights during training for both DINOv2 and DINOv3 object detection models. The change is controlled via a new `backbone_freeze` argument, and includes updates to the training logic and new tests to ensure correct behavior.

**Backbone freezing functionality:**

* Added a `freeze_backbone` method to both `DINOv2LTDETRObjectDetectionTaskModel`, `DINOv3LTDETRObjectDetectionTaskModel` and `PicoDetObjectDetectionTaskModel` which sets the backbone to evaluation mode and disables gradient computation for its parameters.
* Introduced the `backbone_freeze` argument to `DINOv2LTDETRObjectDetectionTrainArgs`, `DINOv3LTDETRObjectDetectionTrainArgs` and `PicoDetObjectDetectionTrainArgs` allowing users to specify whether the backbone should be frozen during training. 
* Modified the `set_train_mode` method in both train model classes to call `freeze_backbone` if `backbone_freeze` is set to `True`. 

**Code quality improvements**

* Standardized the backbone freezing for all models that supports backbone freezing using PyTorch current recommended approach, using Module.requires_grad_

## How has it been tested?

* Added tests in `test_task_model.py` for both DINOv2 and DINOv3 to verify that the backbone freezing logic works as expected, including parameterization for both frozen and unfrozen cases.

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [x] Yes
- [ ] Not needed (internal change)

## Did you update the documentation?

- [x] Yes
- [ ] Not needed (internal change without effects for user)

### CODE REVIEW

File: src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/task_model.py [hunk 1]
```diff
@@ -395,5 +395,4 @@ def load_train_state_dict(self, state_dict: dict[str, Any]) -> None:
 
     def freeze_backbone(self) -> None:
         self.backbone.eval()
-        for param in self.backbone.parameters():
-            param.requires_grad = False
+        self.backbone.requires_grad_(False)
```
Comment (guarin) [inline src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/task_model.py:397]: Thanks for the improvement!

**Comment by gabrielfruet** [comment]
Sure! I can do it.

Just one question that i would like to know your opinion.

Is it desirable to replicate the freezing behavior in two places? shouldn't we keep this either on the TrainModel or TaskModel? Other than BC, I don't know what value we get from adding this to the TaskModel. 

It makes more sense to me the train model being responsible for this, since freezing or unfreezing is related to the training, not task.

I can do either, just let me know what do you think!

File: src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/task_model.py (hunk 1 above)
Comment (gabrielfruet) [reply src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/task_model.py:397]: Thank you!

**Comment by gabrielfruet** [comment]
@liopeer I saw your reaction to my message. Did you mean we should go forward with adding freezing to each TaskModel, or should we leave it out?

**Comment by guarin** [comment]
> @liopeer I saw your reaction to my message. Did you mean we should go forward with adding freezing to each TaskModel, or should we leave it out?

I would prefer to keep freezing for each task model individually. There are always exceptions where the default will not work well and keeping it in the individual task model forces you to adjust it when implementing the model. As an example, we could also add backbone freezing to EoMT, but EoMT injects query parameters in the middle of the backbone. So backbone freezing for EoMT should actually only freeze the layers before the query injection.

> Is it desirable to replicate the freezing behavior in two places? shouldn't we keep this either on the TrainModel or TaskModel? Other than BC, I don't know what value we get from adding this to the TaskModel.

Yes this is desirable, the idea is that the TaskModel controls the model structure and the TrainModel controls the training (model + metrics + loss + train/eval). Both have to know about freezing. Also, people can fine-tune the TaskModel in their own pipeline so it makes sense to expose backbone freezing from the TaskModel as well.

**Comment by gabrielfruet** [comment]
Clear now, thanks for the patience! I’ll get this implemented too.

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py [hunk 2]
```diff
@@ -320,6 +320,7 @@ def __init__(
         classes: dict[int, str],
         image_size: tuple[int, int],
         image_normalize: dict[str, Any] | None = None,
+        backbone_freeze: bool = True,
```
Comment (guarin) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py]: This should be False by default because until now it was not possible to freeze the backbone

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py [hunk 3]
```diff
@@ -730,6 +739,7 @@ def __init__(
...
```
Comment (guarin) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py]: Same here

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py [hunk 4]
```diff
@@ -451,6 +451,7 @@ def __init__(
         classes: dict[int, str],
         image_size: tuple[int, int],
         image_normalize: dict[str, Any] | None = None,
+        backbone_freeze: bool = True,
```
Comment (guarin) [inline src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py]: Same here

File: src/lightly_train/_task_models/semantic_segmentation_multihead/task_model.py [hunk 5]
```diff
@@ -55,15 +55,15 @@ def __init__(
...
+                 List of head names. One segmentation head is created for each name.
                 Head names should follow the format `head_lr{value}` where value is the
                 learning rate formatted without trailing zeros and dots replaced with
                 underscores (e.g., "head_lr0_001" for lr=0.001).
             class_ignore_index:
                 The class index to ignore during training. If provided, an additional
                 class output is created for this ignored class.
             image_size:
-                The size of the input images as (height, width). Default (518, 518).
+                The size of the input images.
```
Comment (guarin) [inline src/lightly_train/_task_models/semantic_segmentation_multihead/task_model.py]: Not sure why this changed?

**Review by guarin — CHANGES_REQUESTED** [review]

File: src/lightly_train/_task_models/semantic_segmentation_multihead/task_model.py (hunk 5 above)
Comment (gabrielfruet) [reply src/lightly_train/_task_models/semantic_segmentation_multihead/task_model.py]: I think it was an erroneous tab-autocomplete

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py (hunk 2 above)
Comment (gabrielfruet) [reply src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py]: Great catch! leaving that `True` was a slip here.

File: src/lightly_train/_task_models/image_classification_multihead/train_model.py [hunk 6]
```diff
@@ -455,6 +455,7 @@ def validation_step(
 
     def set_train_mode(self) -> None:
         self.train()
+        # backbone is always frozen for multihead training
```
Comment (gabrielfruet) [inline src/lightly_train/_task_models/image_classification_multihead/train_model.py]: Added a comment to clarify why multi-head is frozen by default.

**Comment by guarin** [comment]
PR looks good! If you can resolve the merge conflicts we can go ahead and merge it

**Comment by gabrielfruet** [comment]
@guarin Done!

**Comment by straint1** [comment]
Hi, I've been training an object detection model with LT-DETR and DinoV3 backend with the freeze_backbone option. It appears to have broken the validation functionality, giving small MAP scores (0.001). The MAP scores are fine when finetuning the full model. It could very well be a mistake I've made but thought I would sense check this with you.

**Comment by gabrielfruet** [comment]
Hi @straint1 , can you provide the code snippet where you're facing this issue? I'll take a look. Also, if you're using a public dataset, let me know!

**Comment by straint1** [comment]
> Hi @straint1 , can you provide the code snippet where you're facing this issue? I'll take a look. Also, if you're using a public dataset, let me know!

I am running the below - unfortunately the dataset is not publically available, but works fine with ultralytics and with lightly-train before I reinstalled this afternoon following this PR. 

```
names = {int(num) - 1: class_name for (num, class_name) in class_mapping.items()}

    # Fine-tune DINOv3 with DETR for object detection.
    lightly_train.train_object_detection(
        out=out_path,
        overwrite=True,
        model="dinov3/vits16-ltdetr-coco",
        data={
            "path": "/home/ec2-user/data/Maritime_Dataset",
            "train": "images/train",
            "val": "images/val",
            "names": names,
        },
        batch_size=16,
        steps=steps,
        model_args={"backbone_freeze": True},
    )
```

Screenshot 2026-02-24 195518

Here is what is outputted, the training loss decreases as before, (much smoother with the backend frozen btw) but the validation loss is huge (not shown) and low MAP scores. The validation and training sets are from the same dataset so shouldn't diverge by an order of magnitude.

**Comment by guarin** [comment]
Interesting, I haven't observed this. I just tested it with a `dinov3/vits16-ltdetr` model (not pretrained on COCO) and it converged quite nicely:

Screenshot 2026-02-25 at 08 30 25
